### PR TITLE
Contains function supports now regex as second argument 

### LINKED ISSUE #1554: Enhancement : a contains() that support regex
Issue #1550 shows that a [`contains()`](https://github.com/OpenRefine/OpenRefine/blob/master/main/src/com/google/refine/expr/functions/strings/Contains.java) function that accepts Regex might be useful. Now that @jackyq2015 has coded the [`find()`](https://github.com/OpenRefine/OpenRefine/blob/9ab897cb72111de58956721e46fb28f26f121f3b/main/src/com/google/refine/expr/functions/strings/Find.java) function, looks like even a complete beginner would be able to draw on his code to enrich `contains()`. Could it make a "good first issue" ?

### PR DESCRIPTION
**Pull request by gobertm**
Closes #1554 

Implementation based on Partition function.

### CODE REVIEW

**Comment by ettorerizza** [comment]
Looks like special characters such as `. ^ $` are always matched, even in string mode. 

screenshot-127 0 0 1-3333-2018 04 05-12-23-43

Could you try with : 

```
value.contains("[a]")
value.contains("$")
value.contains(".")
value.contains("^")
```

File: main/src/com/google/refine/expr/functions/strings/Contains.java [hunk 1]
```diff
@@ -47,8 +49,14 @@ public Object call(Properties bindings, Object[] args) {
         if (args.length == 2) {
             Object s1 = args[0];
             Object s2 = args[1];
-            if (s1 != null && s2 != null && s1 instanceof String && s2 instanceof String) {
-                return ((String) s1).indexOf((String) s2) > -1;
+            if (s1 != null && s2 != null && s1 instanceof String && (s2 instanceof String || s2 instanceof Pattern)) {
```
Comment (ettorerizza) [inline main/src/com/google/refine/expr/functions/strings/Contains.java]: Do a better distinction between String and Pattern, as in the [replace() method](https://github.com/OpenRefine/OpenRefine/blob/master/main/src/com/google/refine/expr/functions/strings/Replace.java) ?

**Comment by gobertm** [comment]
@ettorerizza It depends what is considered string mode and regexp mode. 
In OpenRefine do you accept regular expression inside double quotes or regular expression must only be in / /  ? 
Because it seems that currently the behavior is not consistent throughout all functions : 
- `find`  (as you also mentioned [here](https://github.com/OpenRefine/OpenRefine/pull/1432#issuecomment-378906943 ). 
but also `match`
 **interprets regular expression in double quotes**. 

image

- `replace` and `partition`  **only consider regexp when in ` / / `**

My original commit of `contains` this morning worked like replace and partition. That's why I couldn't write tests easily (regexp can not be passed as String).

I guess you have to choose between the two interpretation.

**Comment by gobertm** [comment]
@thadguidry 
I just checked all [String functions](https://github.com/OpenRefine/OpenRefine/wiki/GREL-String-Functions) mentionning regular expression.
Those that accepts regex in `" "` are : find and match . Others are ok.

@ettorerizza They indeed act like that currently. If I only keep the initial commit of this PR the contains function works like partition.

**Comment by ettorerizza** [comment]
`Those that accepts regex in " " are : find and match . Others are ok.`

Good news, match() can only be used with regexp anyway. Just leave it like that, it will avoid retrocompatibility problems.

List of functions that use regexp:

```
find()
contains()
match()

partition()
rpartition()
split()
replace()
```

**Comment by ettorerizza** [comment]
@gobertm I'm really not a specialist in Git or Github, but I think it's better to do PRs from a new branch of your master named eg `issue/<issueenumber>`.

**Comment by ettorerizza** [comment]
I think it's okay now.

```
value = "rose is a rose"

value.contains("\s+")--> false
value.contains("\\s+") -->  false
value.contains(/\s+/)--> true
value.contains("$")  --> false
value.contains(/$/)  --> true
value.contains("rose") --> true
value.contains("r.se") --> false

```

**Comment by gobertm** [comment]
Hi,
Is there something else to do in order to have this one completed?

Cheers,
Maxime

**Comment by jackyq2015** [comment]
@gobertm can you please add a unit test?

**Comment by gobertm** [comment]
Hi,
I added the tests based on @ettorerizza tests. 

But sorry I'm messing with rebase and so on. I try to clean this up. :/

**Comment by gobertm** [comment]
Hi,
I added tests based on @ettorerizza examples.
String input , regular expression pattern injected as string are interpreted as string and for the tests of regular expression I tested by directly innjecting a Pattern.

**Pull request merged**